// @nothing-but/utils@0.3.2/dist/chunk/233FGWLT.js
import { __export } from './5ZKAE4VZ.js';

// src/misc.ts
var misc_exports = {};
__export(misc_exports, {
  chain: () => chain,
  entries: () => entries,
  falseFn: () => falseFn,
  isNonNullable: () => isNonNullable,
  isOfClass: () => isOfClass,
  isPlainObject: () => isPlainObject,
  keys: () => keys,
  noop: () => noop,
  reverseChain: () => reverseChain,
  trueFn: () => trueFn
});
var noop = () => void 0;
var trueFn = () => true;
var falseFn = () => false;
var entries = Object.entries;
var keys = Object.keys;
var isPlainObject = (value) => value && Object.getPrototypeOf(value) === Object.prototype;
var isOfClass = (v, c) => v instanceof c || v && v.constructor === c;
var isNonNullable = (i) => i != null;
function chain(callbacks) {
  return (...args) => {
    for (const callback of callbacks)
      callback && callback(...args);
  };
}
function reverseChain(callbacks) {
  return (...args) => {
    for (let i = callbacks.length - 1; i >= 0; i--) {
      const callback = callbacks[i];
      callback && callback(...args);
    }
  };
}

export { chain, entries, falseFn, isNonNullable, isOfClass, isPlainObject, keys, misc_exports, noop, reverseChain, trueFn };
